### PR TITLE
fix completion token extraction when using chat templates

### PR DESCRIPTION
**Pull request by vlad-karp**
# Description

Whenever apply_chat_template() is called in the input_pipeline (SFT/RL scenario), it was assumed that the tokenizer_model.apply_chat_template() returns a single list of tokens. 
But starting from some HF version (~4.40), it already returns BatchEncoding object if you use it with tokenization enabled.

Then the logic of extracting the completion part of a conversation round appears to be broken.

This PR changes the extraction logic by handling multiple types of tokenizer_model.apply_chat_template() outputs - BatchEncoding, dict and lists(old approach).

This change is required to correctly use chat templates preprocessing for SFT training scenarios when a relatively new HF transformer's library tokenizers is used.

# Tests
python3 -m tests.integration.sft_trainer_correctness_test

This is the supposed test to check this logic, but it is disabled currently. 
I had to regenerate the gold tokens & their logits since the old HF llama2 tokenizer is not correctly applying its chat template.

If I force execution of sft_trainer_correctness_test test by adding
"""
if __name__ == "__main__":
  unittest.main()
"""
To sft_trainer_correctness_test, it passes.
If I run it via pytest it still fails for some reason so I keep it disabled.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 1]
```diff
@@ -171,6 +172,41 @@ def is_conversational(features, data_columns):
...
+    prompt_completion_ids = prompt_completion_tokens.input_ids
+    prompt_ids = prompt_tokens.input_ids
+  elif isinstance(prompt_completion_tokens, dict) and INPUT_TOKENS_KEY in prompt_completion_tokens:
+    prompt_completion_ids = prompt_completion_tokens[INPUT_TOKENS_KEY]
+    prompt_ids = prompt_tokens[INPUT_TOKENS_KEY]
+  elif isinstance(prompt_completion_tokens, list):
+    prompt_completion_ids = prompt_completion_tokens
+    prompt_ids = prompt_tokens
+  else:
+    raise ValueError(f"Can't hadnle the chat template output of type {type(prompt_completion_tokens)}")
```
Comment (entrpn) [inline src/maxtext/input_pipeline/input_pipeline_utils.py]: nit: mispelled handle

File: tests/assets/logits_generation/generate_sft_golden_data.py [hunk 2]
```diff
@@ -199,7 +200,14 @@ def test_with_trl_and_save_golden_data(config):
...
   parser.add_argument("--model-name", type=str, required=False, default="llama2-7b")
-  parser.add_argument("--tokenizer-path", type=str, required=False, default="meta-llama/Llama-2-7b-chat-hf")
+  # In transformers=5.2.0 (at least), the Llama-2 tokenizer incorrectly injects an extra space
+  # before <s> and [INST] tokens when applying its chat template.
+  # Also, using the local tokenizer is consistent with the tokenizer used by sft_trainer_correctness_test.py
+  # which depends on generated gold data here.
+  parser.add_argument(
+      "--tokenizer-path", type=str, required=False, default=os.path.join(MAXTEXT_ASSETS_ROOT, "llama2-chat-tokenizer")
+  )
+  parser.add_argument("--model-path", type=str, required=False, default="meta-llama/Llama-2-7b-chat-hf")
```
Comment (gagika) [inline tests/assets/logits_generation/generate_sft_golden_data.py]: nit: could you have `--hf-model-path` to be consistent with rest of maxtext code?

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 3]
```diff
@@ -171,6 +172,41 @@ def is_conversational(features, data_columns):
...
+  Returns:
+    A string representing the completion formatted by the chat template.
+  """
+  prompt_completion_tokens = tokenizer_model.apply_chat_template(round_msgs, add_generation_prompt=False, tokenize=True)
+  prompt_tokens = tokenizer_model.apply_chat_template(round_msgs[:-1], add_generation_prompt=False, tokenize=True)
+
+  # attention masks in BatchEncoding are effectively ignored
+  if hasattr(prompt_completion_tokens, INPUT_TOKENS_KEY):
+    prompt_completion_ids = prompt_completion_tokens.input_ids
+    prompt_ids = prompt_tokens.input_ids
```
Comment (gagika) [inline src/maxtext/input_pipeline/input_pipeline_utils.py]: as you already are using `INPUT_TOKENS_KEY = "input_ids"` here you can reply only `INPUT_TOKENS_KEY` and have:

    prompt_completion_ids = getattr(prompt_completion_tokens, INPUT_TOKENS_KEY)
    prompt_ids = getattr(prompt_tokens, INPUT_TOKENS_KEY)
